### ISSUE TITLE
GCE LoadBalancer creates FW rules with tags assuming GKE generated instances

### ISSUE THREAD

**Issue by kshelton** [kubernetes/kubernetes#13035]
The GCE LoadBalancer implementation creates fw rules with tags where it assumes that 
* Instances all contain the same prefix followed by a "-" and a random 5 characters
* Instances all contain tags with the prefix

This causes unnecessary constraints on instance names and forces hostname+tags that match this pattern, even in a non-GKE environment.

**Comment by kshelton** [comment]
Parent  3837f8c was against 1.0.3

**Comment by thockin** [comment]
@a-robinson is this resolved now?

**Comment by a-robinson** [comment]
We no longer require the prefix to be followed by a "-" and 5 characters (now it can be a "-" and any number of characters), but we still assume all the nodes are tagged with the prefix. 

@kshelton - what would you propose doing differently? If we don't require tags, then we'll be opening firewalls to all the VMs in the user's network, which could be a real security issue in some cases.

We could lift the requirement on VM names by passing through the cluster name to this code and using that as the tag, but would that really make things easier on you?

**Comment by kshelton** [comment]
@a-robinson So, this patch implies that if there is a tag that matches that is a prefix of the instance name, then that will be added to the fiw rule.  This allows the GKE case of auto tagging instances to still function and gives the user the option to tag their instances if they so desire, which allows this to function in GCE without the GKE naming/tagging convention.  

As for the security concerns, this does not break the current functionality, but I believe that tagging should ultimately be computed more intelligently (e.g. intersection of tags for LB targets) or user provided.  This should allow more flexibility with custom tagging of instances based on services and selectors and lead to greater security.

**Comment by a-robinson** [comment]
Only opening the firewall on certain nodes is likely to cause problems, given that we include all of the nodes in the cluster in the GCE Target Pool. If any of those nodes in the target pool don't have the proper firewall, the fraction of requests that the load balancer forwards to them will get dropped on the floor.

At a higher level, there are two levels of selection here for choosing which nodes have to open firewalls or receive traffic -- which nodes the pods/endpoints that back the service are running on, as well as which nodes are part of the target pool and firewall. Both of these layers would have to be controlled in order to keep a service's traffic off of "untrusted" nodes. We plan to improve how we create target pools to only include the nodes on which the service's endpoints are running. At that point, users will be able to fully control which nodes receive requests for a service all in one natural place -- the scheduling of their pods.

As for tagging, I do think there's value in being more flexible, but the code in https://github.com/kubernetes/kubernetes/commit/3837f8c9d0eba0333ef94a77791587d4a1aafa53
(1) would make an unsustainable amount of GCE API calls for large clusters with a lot of services
(2) will cause nodes without tags to still be in the target pool without having the firewall applied, leading to dropped traffic (as mentioned above)

I could maybe see replacing what we have with the core logic of your code, where we would look up the tags on a single node and pick the tag or tags that are prefixes of the node name. It'd be an ever so slightly more general version of what we currently have, but on the downside it probably wouldn't really help anyone that our current strategy isn't good enough for.

**Comment by kshelton** [comment]
@a-robinson in regards (2) How about only allowing either a) no tags or b) all having matching tags, throwing an exception if there is a mix.  I think the ultimate solution is allowing the user to provide the tag(s) (or defaulting to this method), but I'm am unsure how to plumb that through in a generic way (do the other providers have a tag equivalent?).

In regards to (1), how about changing the tag lookup to use https://cloud.google.com/compute/docs/reference/latest/instances/list with a filter for the instance names and items(name,tags) for the fields?  This should allow retrieving all the tags in one API call.

**Comment by a-robinson** [comment]
Help me understand the use case here and why it's important. From what I see, such a solution's advantages over the current logic are:

1. It's more flexible in that it falls back to using no tags in the case that not all nodes share a tag
1. It can handle a wider variety of prefix-style tags.

However, it still either uses no tags or instance name prefix tags, and incurs extra work in querying GCE for the nodes, when we already have an up-to-date list of them (just without the tag info). There doesn't seem to be the potential for security benefit mentioned above, and that's going to be true so long as all nodes in the cluster are included in the target pool. I'm not sure how useful supporting different instance name prefix tags is, since if you're going to all the trouble to put instance-name-prefix tags on all your VMs, why not just put the same one on all your VMs?

Is there a simpler way to fall back to using no tags if that's what the user wants?

I think the main disconnect here is that I'm assuming that most clusters on GCE are using the well-supported scripts included in the repo, which take care of the tags for you, allowing this code to be simpler and more reliable in the common case, but at the expense of not supporting different tagging schemes. Is that a faulty assumption on my part?

**Comment by kshelton** [comment]
In regards to the security aspect, I disagree.  The tagging from the current master assumes that all the instances are tagged the **same**.  The only security this provides is if you are running other nodes than your cluster in the same project, otherwise traffic would still be allowed to all instances (per the FW rule, not the pool).  

With the proposed change, one could tag all the instances with the exact name of the instance and when the change for only adding the instances that are running the service to the pool is merged - only those instances would be tagged into the firewall rule.

It is not as simple as adding a prefixed tag to the instances.  With the current implementation all the nodes must have the **same** prefix and prefix tag.  The proposed change would add the set of prefix tags, allowing them to be different.  In addition, the fact is that the current method would take a instanced named something like 'etcd-1' and make a tag 'e', which in my opinion is broken.

Imagine a scenario where you are running services that have different needs: more CPU, more RAM, local SSD disks, etc.  You would then have the service use a nodeSelector that matches how you have tagged k8s for those nodes (I use fleet for this and propagate the fleet metadata to k8s ).  By allowing the instance tags to be different, you can restrict the FW rules so that they do not effect any other node with traffic on ports it shouldn't be receiving traffic on.

Currently, or even after you only add the proper nodes to the pool, any instance attached IP address (not sent to a forwarding rule) could receive traffic on any port that a service is exposed on, even if there is no forwarding rule/pool to route the traffic to that node, because of the tagging.

**Comment by a-robinson** [comment]
> In regards to the security aspect, I disagree. The tagging from the current master assumes that all the instances are tagged the same. The only security this provides is if you are running other nodes than your cluster in the same project, otherwise traffic would still be allowed to all instances (per the FW rule, not the pool).
> 
> With the proposed change, one could tag all the instances with the exact name of the instance and when the change for only adding the instances that are running the service to the pool is merged - only those instances would be tagged into the firewall rule.

Yes, I meant that the security is the same at this point in time. Regarding the future, the existing code assumes that all instances *passed to CreateTCPLoadBalancer* are tagged the same. That will automatically adapt to us only passing CreateTCPLoadBalancer the hosts with a service's endpoints on them, enabling cases where pools of machines with different properties have different names and tags.

The main security benefit will be the case you mention, where each instance has its own unique tag. There is some real potential value here if the user has put k8s scheduling constraints on the nodes without creating pools of names/tags in GCE.

> It is not as simple as adding a prefixed tag to the instances. With the current implementation all the nodes must have the same prefix and prefix tag. The proposed change would add the set of prefix tags, allowing them to be different. In addition, the fact is that the current method would take a instanced named something like 'etcd-1' and make a tag 'e', which in my opinion is broken.

It would make the tag 'etcd', not 'e', as of the change merged into master three weeks ago(https://github.com/kubernetes/kubernetes/pull/12438) and into release-1.0 last week (https://github.com/kubernetes/kubernetes/pull/13127).

> Currently, or even after you only add the proper nodes to the pool, any instance attached IP address (not sent to a forwarding rule) could receive traffic on any port that a service is exposed on, even if there is no forwarding rule/pool to route the traffic to that node, because of the tagging.

There was some work on this (https://github.com/kubernetes/kubernetes/pull/12088), but it may have been abandoned. cc @ArtfulCoder 


Thanks for explaining further -- consider me interested. What I'd like to do is structure this in a way that doesn't add much overhead to the process. I'll come back to this in an hour or two with more thoughts, but have to run now.

**Comment by a-robinson** [comment]
We may be able to get by with just a simple cache for the host -> tag mappings, but tags on GCE can be changed while the VM is running so cluster admins would be able to shoot themselves in the foot. They can already shoot themselves in the foot in a variety of ways though, so perhaps that's fine.

**Comment by kshelton** [comment]
>Regarding the future, the existing code assumes that all instances passed to CreateTCPLoadBalancer are tagged the same.
Good point, I was thinking of the GKE case where the are all tagged the same, but that's not enforced by the k8s code.

So, I think having the constraint of "If you change your instance tags you need to recreate your LB and restart the api servers / kubelets" seems reasonable.  For a path forward then, are you fine with keeping the generation to use the set of tags that are the instance name prefixes (up to including the full name) and just caching the set of host->tag mappings?  Perhaps we can async cache all hosts->tags for the project and expire them (blocking if cache isn't populated yet) and change the query over to the bulk fetch/

**Comment by a-robinson** [comment]
Sure, I think the instance prefix approach with a cache for the host->tag mappings is workable. Feel free to send a PR my way.

**Comment by kshelton** [comment]
As soon as I get the rest of my infrastructure deployed I'll go back and update the code as discussed.  The hack works for my immediate concern right now, so I need to get this stuff out there before I have time to fix it.  I'll make a PR as soon as I get it done

**Comment by kshelton** [comment]
Implemented without a cache because forwarding rules, etc. are already being retrieved without caching.  

I think as a next improvement there should probably be a more generic GCE level cache like the kubernetes cache in the service controller.  I looked at using that implementation, but it is specific to the format of the kubernetes api.  Perhaps it would be best to generalize that component, as it already has an interface for listening for changes.
